src/model/wallet.js

```diff
@@ -222,24 +238,187 @@ export const loadAddress = async () => {
+    if (!isSeedPhraseMigrated) {
+      const { privateKey } = await migrateSecrets();
+      return privateKey;
+    } else {
+      const address = await loadAddress();
+      const { privateKey } = getPrivateKey(address, authenticationPrompt);
+      return privateKey;
+    }
+  } catch (error) {
+    captureException(error);
+    return null;
+  }
+};
+
+export const saveAddress = async (address, accessControlOptions = {}) => {
+  await keychain.saveString(addressKey, address, accessControlOptions);
+};
+
+export const identifyWalletType = walletSeed => {
+  let type = null;
+  if (
+    isHexStringIgnorePrefix(walletSeed) &&
+    addHexPrefix(walletSeed).length === 66
+  ) {
+    type = WalletTypes.privateKey;
+    // 12 or 24 words seed phrase
+  } else if (isValidMnemonic(walletSeed)) {
+    type = WalletTypes.mnemonic;
+    // Public address (0x)
+  } else if (isValidAddress(walletSeed)) {
+    type = WalletTypes.readOnly;
+  } else {
+    // seed
+    type = WalletTypes.seed;
+  }
+
+  return type;
+};
+
+export const newCreateWallet = async (seed, color, name) => {
+  const isImported = !!seed;
   const walletSeed = seed || generateSeedPhrase();
   let wallet = null;
+  let hdnode = null;
+  let addresses = [];
   try {
-    if (
-      isHexStringIgnorePrefix(walletSeed) &&
-      addHexPrefix(walletSeed).length === 66
-    ) {
-      wallet = new ethers.Wallet(walletSeed);
-    } else if (isValidMnemonic(walletSeed)) {
-      wallet = ethers.Wallet.fromMnemonic(walletSeed);
-    } else {
-      let hdnode = ethers.utils.HDNode.fromSeed(walletSeed);
-      let node = hdnode.derivePath("m/44'/60'/0'/0/0");
+    const type = identifyWalletType(walletSeed);
+    if (!type) throw new Error('Unknown Wallet Type');
+    switch (type) {
+      case WalletTypes.privateKey:
+        wallet = new ethers.Wallet(walletSeed);
+        break;
+      case WalletTypes.mnemonic:
+        hdnode = ethers.utils.HDNode.fromMnemonic(walletSeed);
+        break;
+      case WalletTypes.seed:
+        hdnode = ethers.utils.HDNode.fromSeed(walletSeed);
+        break;
+      case WalletTypes.readOnly:
+        wallet = { address: toChecksumAddress(walletSeed), privateKey: null };
+        break;
+      default:
+    }
+
+    // Always generate the first account
+    if (!wallet && [WalletTypes.mnemonic, WalletTypes.seed].includes(type)) {
+      const node = hdnode.derivePath(`${DEFAULT_HD_PATH}/0`);
       wallet = new ethers.Wallet(node.privateKey);
     }
+
+    // Get all wallets
+    const allWalletsResult = await getAllWallets();
+    const allWallets = get(allWalletsResult, 'wallets', {});
+
+    if (isImported) {
+      // Checking if the generated account already exists
+      const alreadyExisting = Object.keys(allWallets).some(key => {
+        const someWallet = allWallets[key];
+        return someWallet.addresses.some(account => {
+          return account.address === wallet.address && account.type === type;
```

Comment:
are the addresses from ethers wallet object always checksummed?

Reply:
Yes

Reply:
I'll add it anyway cause read only wallets don't go through ethers